streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -209,15 +206,29 @@ private void addNewTask(final Task task) {
      * @throws StreamsException if the store's change log does not contain the partition
      */
     boolean checkForCompletedRestoration() {
+        boolean allRunning = true;
+
+        // first initialize the created tasks, then check if they can complete the restoration
         final List<Task> restoringTasks = new LinkedList<>();
         for (final Task task : tasks.values()) {
+            if (task.state() == CREATED) {
+                try {
+                    task.initializeIfNeeded();
+                } catch (final LockException e) {
+                    // it is possible that if there are multiple threads within the instance that one thread
+                    // trying to grab the task from the other, while the other has not released the lock since
+                    // it did not participate in the rebalance. In this case we can just retry in the next iteration
+                    log.debug("Could not initialize {} due to {}; will retry", task.id(), e.toString());
+                    allRunning = false;
+                }
+            }
+
             if (task.state() == RESTORING) {
                 restoringTasks.add(task);
             }
         }
 
-        boolean allRunning = true;
-        if (!restoringTasks.isEmpty()) {
+        if (allRunning && !restoringTasks.isEmpty()) {
```

Comment:
Not sure this conditional makes sense... It says "if all the tasks are running, and there are restoring tasks", which is self-contradictory. Maybe we can have two boolean flags, `all initialized` and `all running`?

Reply:
The rationale is that some tasks may still be in CREATED state and some maybe in RESTORING, in either case we should set `allRunning` to false. If some are still in CREATED state then `allRunning` would be false here already so the first condition is just to save the second check.

So I think correctness-wise that is fine but maybe a bit confusing, I can make this logic a bit cleaner now.